packages/noya-state/src/reducers/history.ts

```diff
@@ -0,0 +1,99 @@
+import produce from 'immer';
+import { ApplicationState, Action, reducer } from './application';
+import { createInitialInteractionState } from './interaction';
+
+export type HistoryEntry = {
+  actionType: Action[0];
+  timestamp: number;
+  state: ApplicationState;
+};
+
+export type HistoryState = {
+  past: HistoryEntry[];
+  present: ApplicationState;
+  future: HistoryEntry[];
+};
+
+export type HistoryAction = [type: 'undo'] | [type: 'redo'] | Action;
+
+const FILE_CHANGED_TIMEOUT = 300;
+
+export function historyReducer(state: HistoryState, action: HistoryAction) {
+  const currentPresent = state.present;
+  const [actionType] = action;
+  switch (actionType) {
```

Comment:
We should do `switch (action[0]) { ... }` here. Defining `actionType` as a separate var will prevent typescript from narrowing the type of `action`

Reply:
Ah, I was wondering why it was complaining 😅